### PR TITLE
Setup backend to save teacher apps

### PR DESCRIPTION
**Pull request by megcrenshaw**
Together with the frontend https://github.com/code-dot-org/code-dot-org/pull/44441, this fulfills _Req 1: Add a “Save and Return Later” Button to the application_. See below for future work.

Current functionality: https://watch.screencastify.com/v/SqvvikhJJq8mzL4Ygauv 

I did the following:
- (Backend): Ignore validations if the application is incomplete.
- (Backend): Manually set some fields on an update, and only do some actions on a completed application.

## Links

- spec: [Saving Progress and Reopening Applications](https://docs.google.com/document/d/1kbRAQ3To-MHVmz2KHnyaN6Mh6EiJk8F_zRNS1F8kSK8/edit?pli=1#heading=h.86qvf1hodt)
- jira ticket: [PLAT-1466](https://codedotorg.atlassian.net/browse/PLAT-1466)

## Follow-up work

In (a) separate PR(s), I'm planning to
- address the to-dos,
- address two `assert_nil` warnings in relevant backend tests, and

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb [hunk 1]
```diff
@@ -13,14 +13,18 @@ def new_form
 
     # PATCH /api/v1/pd/application/teacher/<applicationId>
     def update
-      form_data_hash = params.try(:[], :form_data) || {}
-      form_data_json = form_data_hash.to_unsafe_h.to_json.strip_utf8mb4
+      form_data_hash = params.try(:[], :form_data)
+      form_data_json = form_data_hash.to_unsafe_h.to_json.strip_utf8mb4 if form_data_hash || '{}'
```
Comment (davidsbailey) [inline dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb]: this doesn't look right -- since the `||` takes precedence over the trailing `if`, it looks like the condition after the `if` will always be truthy.
```
[development] dashboard > 'foo' if false || 'bar'
=> "foo"
```

File: dashboard/app/models/pd/application/teacher_application.rb [hunk 2]
```diff
@@ -102,10 +102,12 @@ class TeacherApplication < ApplicationBase
...
     # 1. Application has a specific school? always overwrite the user's school info
     # 2. User doesn't have a specific school? overwrite with the custom school info.
+    # Will only update the school info if we have enough information for it
+    # An incomplete application may not have all the information we need to update
     def update_user_school_info!
       if school_id || user.school_info.try(&:school).nil?
-        school_info = get_duplicate_school_info(school_info_attr) || SchoolInfo.create!(school_info_attr)
-        user.update_school_info(school_info)
+        school_info = school_info_attr
+        user.update_school_info(get_duplicate_school_info(school_info) || SchoolInfo.create!(school_info)) if school_info
```
Comment (davidsbailey) [inline dashboard/app/models/pd/application/teacher_application.rb]: it looks like there is already some test coverage for this method. could you please add a new test to cover the new behavior?

File: dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb [hunk 3]
```diff
@@ -13,14 +13,18 @@ def new_form
...
 
-      # [MEG] TODO: See if update is coming from hitting the submit button (do validations)
-      # or the save button (ignore validations)
       @application.form_data_hash = JSON.parse(form_data_json)
+      @application.set_status
+      @application.set_course_from_program
+      @application.update_user_school_info!
```
Comment (davidsbailey) [inline dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb:21]: this pattern looks fragile because it creates a hidden dependency in the order in which these methods are called. I would recommend doing this in a more declarative manner:
```
hash = ...
set_status(hash[:status])
...
```

File: dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb [hunk 4]
```diff
@@ -13,14 +13,18 @@ def new_form
...
-      form_data_hash = params.try(:[], :form_data) || {}
-      form_data_json = form_data_hash.to_unsafe_h.to_json.strip_utf8mb4
+      form_data_hash = params.try(:[], :form_data)
+      form_data_json = form_data_hash.to_unsafe_h.to_json.strip_utf8mb4 if form_data_hash || '{}'
 
-      # [MEG] TODO: See if update is coming from hitting the submit button (do validations)
-      # or the save button (ignore validations)
       @application.form_data_hash = JSON.parse(form_data_json)
+      @application.set_status
+      @application.set_course_from_program
```
Comment (davidsbailey) [inline dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb]: why is this necessary if it will also be done `before_validation`?

File: dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb (hunk 1 above)
Comment (megcrenshaw) [reply dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb]: Yikes! This is what I get for making a last-second change. Will update ...

File: dashboard/app/models/pd/application/teacher_application.rb (hunk 2 above)
Comment (megcrenshaw) [reply dashboard/app/models/pd/application/teacher_application.rb]: Good call, will do

Comment (davidsbailey) [inline dashboard/app/models/pd/application/teacher_application.rb]: sorry to go into the weeds on style here, but I think this logic is getting a bit complicated for a trailing if. a pattern that's worked great for us in ruby for small functions like this one is to prefer early returns over conditionals and end up with something like this:
```suggestion
      return unless school_id || user.school_info.try(&:school).nil?
      school_info = school_info_attr
      return unless school_info
      user.update_school_info(get_duplicate_school_info(school_info) || SchoolInfo.create!(school_info))
```
when there is only one layer of `if` logic, early returns are about the same as conditionals. but the nice part is that this approach completely keeps you out of nested conditionals. it doesn't work great for large methods, but that can be a benefit too, because it forces you to to break up large methods into smaller ones.

of course there's no code.org style rule against conditionals or nested conditionals... and you could also just change the trailing if to
```
      if school_info
        user.update_school_info(get_duplicate_school_info(school_info) || SchoolInfo.create!(school_info))
      end
```
my reason for pushing on getting rid of the trailing `if` in this situation is that once you have a compound statement (due to the `||`) it gets harder to reason about what the trailing `if` is doing.

Comment (megcrenshaw) [reply dashboard/app/models/pd/application/teacher_application.rb]: Ah, this is very helpful––I'm not familiar with best style in Ruby, so this is appreciated. Will refactor!

File: dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb (hunk 4 above)
Comment (megcrenshaw) [reply dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb]: Good call. I was originally not doing `set_course_from_program` `before_validation` if program was nil, but that's not necessary now –– thanks!

File: dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb (hunk 3 above)
Comment (megcrenshaw) [reply dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb:21]: I think ideally, we'd set the form_data_hash, and then a save from there would update the status and the school info? Does that sound right?

File: dashboard/app/models/pd/application/application_base.rb [hunk 5]
```diff
@@ -110,6 +110,7 @@ class ApplicationBase < ApplicationRecord
     before_create -> {self.status = :unreviewed}
     after_initialize :set_type_and_year
     before_validation :set_type_and_year
+    validate :status_is_valid_for_application_type
```
Comment (davidsbailey) [inline dashboard/app/models/pd/application/application_base.rb:113]: just curious if you moved this for a particular reason, e.g. does this change the order in which the validations run in a way that is important? I also noticed that it looked like this had some comments attached to it at its old location, so if you do want to move it, then those should come with it.

File: dashboard/app/models/pd/application/teacher_application.rb [hunk 6]
```diff
@@ -84,12 +84,12 @@ class TeacherApplication < ApplicationBase
...
     has_many :emails, class_name: 'Pd::Application::Email', foreign_key: 'pd_application_id'
 
+    before_validation :set_course_from_program
     validates :status, exclusion: {in: ['interview'], message: '%{value} is reserved for facilitator applications.'}
-    validates :course, presence: true, inclusion: {in: VALID_COURSES}, unless: -> {status == 'incomplete'}
     validate :workshop_present_if_required_for_status, if: -> {status_changed?}
 
-    before_validation :set_course_from_program, unless: -> {program.nil?}
     before_save :save_partner, if: -> {form_data_changed? && regional_partner_id.nil? && !deleted?}
+    before_save :course, presence: true, inclusion: {in: VALID_COURSES}, unless: -> {status == 'incomplete'}
```
Comment (davidsbailey) [inline dashboard/app/models/pd/application/teacher_application.rb:92]: why is this changing to before_save? it looks more like a validation (what it was before). if you need to change this to get things working, let's discuss.

File: dashboard/app/models/pd/application/teacher_application.rb [hunk 7]
```diff
@@ -1146,12 +1156,14 @@ def total_score
...
     end
 
-    # Called after the application is created. Do any manipulation needed for the form data
-    # hash here, as well as send emails
-    # [MEG] TODO: should only do a lot of this on a completed submitted application
     def on_successful_create
       update_user_school_info!
-      queue_email :confirmation, deliver_now: true unless status == 'incomplete'
+      on_completed_application unless status == 'incomplete'
+      save
```
Comment (davidsbailey) [inline dashboard/app/models/pd/application/teacher_application.rb:1162]: I would prefer `save!` to catch if any validations failed

File: dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb (hunk 3 above)
Comment (davidsbailey) [reply dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb:21]: I'm open to keeping it this way but I'm not sure I follow. I have a few other questions about this PR so if it's ok could we meet this afternoon to discuss?

Comment (megcrenshaw) [reply dashboard/app/controllers/api/v1/pd/application/teacher_applications_controller.rb:21]: Yes––I have some questions I'd love to run by you. Thank you!

**Comment by megcrenshaw** [comment]
Doing a refactor elsewhere before coming back to this ... closing for now
